cranelift/codegen/src/egraph/cost.rs

```diff
@@ -43,12 +86,49 @@ impl Cost {
         Cost(0)
     }
 
+    fn new(opcode_cost: u32, depth: u8) -> Cost {
+        debug_assert!(
+            opcode_cost <= Self::MAX_OP_COST,
+            "Cost::new: given opcode cost of {opcode_cost} is larger than max of {}",
+            Self::MAX_OP_COST,
+        );
+        Cost((opcode_cost << Self::DEPTH_BITS) | u32::from(depth))
+    }
+
+    fn depth(&self) -> u8 {
+        let depth = self.0 & Self::DEPTH_MASK;
+        u8::try_from(depth).unwrap()
+    }
+
+    fn op_cost(&self) -> u32 {
+        (self.0 & Self::OP_COST_MASK) >> Self::DEPTH_BITS
+    }
+
     /// Clamp this cost at a "finite" value. Can be used in
     /// conjunction with saturating ops to avoid saturating into
     /// `infinity()`.
     fn finite(self) -> Cost {
         Cost(std::cmp::min(u32::MAX - 1, self.0))
     }
+
+    /// Compute the cost of the operation and its given operands.
+    ///
+    /// Caller is responsible for checking that the opcode came from an instruction
+    /// that satisfies `inst_predicates::is_pure_for_egraph()`.
+    pub(crate) fn of_pure_op(op: Opcode, operand_costs: impl IntoIterator<Item = Self>) -> Self {
+        let c = pure_op_cost(op) + operand_costs.into_iter().sum();
+        Cost::new(c.op_cost(), c.depth().saturating_add(1)).finite()
```

Comment:
Should we only add `1` to depth when the `operand_costs` was non-empty?

Reply:
I think it is fine to have `depth == 1` for ops without operands, and only adding `1` when we have non-empty operands requires avoiding `sum()` and makes the method look less elegant so I'm not gonna do it.